### PR TITLE
ADEN-5712 | Integrate A9

### PR DESCRIPTION
**Pull request by kvas-damian**
https://wikia-inc.atlassian.net/browse/ADEN-5712

wgCountry: `wgAdDriverA9BidderCountries`

Use `apstag.debug('enable');` to force test bid

### CODE REVIEW

File: extensions/wikia/AnalyticsEngine/AnalyticsProviderAmazonMatch.php [hunk 1]
```diff
@@ -1,5 +1,6 @@
 <?php
 
+// TODO ADEN-5756 Remove AmazonMatch after A9 full rollout
```
Comment (WikiaTech) [inline extensions/wikia/AnalyticsEngine/AnalyticsProviderAmazonMatch.php:3]: INFO Complete the task associated to this "TODO" comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1135)

File: extensions/wikia/AdEngine/js/run/mercury.run.js [hunk 2]
```diff
@@ -53,7 +59,13 @@ require([
 	mercuryListener.onLoad(function () {
 		var isRubiconFastlanePrebidAdapterActive = adContext.getContext().bidders.fastlane;
 
-		if (geo.isProperGeo(instantGlobals.wgAmazonMatchCountriesMobile)) {
+		if (geo.isProperGeo(instantGlobals.wgAdDriverA9BidderCountries)) {
+			a9.call();
+		}
+
+		// TODO ADEN-5756 remove 'if' after A9 full roll out
```
Comment (WikiaTech) [inline extensions/wikia/AdEngine/js/run/mercury.run.js:66]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: extensions/wikia/AdEngine/js/AdEngineRunner.js [hunk 3]
```diff
@@ -6,14 +6,15 @@ define('ext.wikia.adEngine.adEngineRunner', [
 	'wikia.log',
 	'wikia.window',
 	require.optional('ext.wikia.adEngine.lookup.amazonMatch'),
+	require.optional('ext.wikia.adEngine.lookup.a9'),
```
Comment (fraszczakszymon) [inline extensions/wikia/AdEngine/js/AdEngineRunner.js]: Can we move it one line above? 😊

File: extensions/wikia/AdEngine/js/lookup/a9.js [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	'wikia.document',
+	'wikia.log',
+	'wikia.window'
+], function (slotsContext, factory, doc, log, win) {
+	'use strict';
+
+	var logGroup = 'ext.wikia.adEngine.lookup.a9',
+		config = {
+			oasis: {
+				BOTTOM_LEADERBOARD: [[728, 90], [970, 250]],
```
Comment (fraszczakszymon) [inline extensions/wikia/AdEngine/js/lookup/a9.js]: Should we handle other slots?

File: extensions/wikia/AdEngine/js/lookup/a9.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+		return a9Slots;
+	}
+
+	function configureApstagCommand(command, args) {
+		win.apstag._Q.push([command, args]);
+	}
+
+	function configureApstag() {
+		if (typeof win.apstag === 'undefined') {
```
Comment (fraszczakszymon) [inline extensions/wikia/AdEngine/js/lookup/a9.js]: What about `win.apstag = win.apstag || {}`?

File: extensions/wikia/AdEngine/js/lookup/a9.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+			win.apstag = {};
+		}
+
+		if (typeof win.apstag._Q === 'undefined') {
```
Comment (fraszczakszymon) [inline extensions/wikia/AdEngine/js/lookup/a9.js]: What about `win.apstag._Q = win.apstag._Q || []`

File: extensions/wikia/AdEngine/js/lookup/a9.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+				configureApstagCommand('f', arguments);
+			};
+		}
+	}
+
+	function calculatePrices() {
+		log(['calculatePrices', bids], 'debug', logGroup);
+
+		Object.keys(bids).forEach(function(slotName) {
+			priceMap[slotName] = bids[slotName].amznbid;
```
Comment (fraszczakszymon) [inline extensions/wikia/AdEngine/js/lookup/a9.js:114]: Should we reset priceMap state on consecutive pages on Mercury?

File: extensions/wikia/AdEngine/js/lookup/a9.js [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+		var bid = bids[slotName];
+
+		log(['getSlotParams', slotName], 'debug', logGroup);
+
+		if (!bid) {
+			return {};
+		}
+
+		return {
+			amznbid: bid.amznbid,
```
Comment (fraszczakszymon) [inline extensions/wikia/AdEngine/js/lookup/a9.js:132]: Is there anything else in this object? Is it worth to rewrite it?

File: extensions/wikia/AdEngine/js/lookup/a9.js [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+			loaded = true;
+		}
+
+		slots = slotsContext.filterSlotMap(config[skin]);
+		a9Slots = getA9Slots(slots);
+		log(['call - fetchBids', a9Slots], 'debug', logGroup);
+
+		win.apstag.fetchBids({
+			slots: a9Slots,
```
Comment (fraszczakszymon) [inline extensions/wikia/AdEngine/js/lookup/a9.js:62]: What's happening with already created slots on consecutive pages on Mercury?

File: extensions/wikia/AdEngine/js/lookup/a9.js [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+			timeout: 2000
+		}, function(currentBids) {
+			log(['call - fetchBids response', currentBids], 'debug', logGroup);
+
+			currentBids.forEach(function (bid) {
+				bids[bid.slotID] = bid;
```
Comment (fraszczakszymon) [inline extensions/wikia/AdEngine/js/lookup/a9.js:68]: Should we reset bids state on consecutive pages on Mercury?

Comment (fraszczakszymon) [inline extensions/wikia/AdEngine/js/lookup/a9.js]: Should we reset bids state on consecutive pages on Mercury?

File: extensions/wikia/AdEngine/js/lookup/a9.js (hunk 8 above)
Comment (kvas-damian) [reply extensions/wikia/AdEngine/js/lookup/a9.js:132]: Yes, there are more attributes in this object (like slotID, mediaType, queryParams). I don't think we should send all of them to DFP.

File: extensions/wikia/AdEngine/js/lookup/a9.js (hunk 9 above)
Comment (kvas-damian) [reply extensions/wikia/AdEngine/js/lookup/a9.js:62]: This method recreates `a9Slots` on every call, so should be ok for Mercury.

File: extensions/wikia/AdEngine/js/lookup/a9.js (hunk 10 above)
Comment (kvas-damian) [reply extensions/wikia/AdEngine/js/lookup/a9.js:68]: Done in https://github.com/Wikia/app/pull/13603/commits/2d523760307796875dadae429ac8bc34e9f5533b

File: extensions/wikia/AdEngine/js/lookup/a9.js (hunk 7 above)
Comment (kvas-damian) [reply extensions/wikia/AdEngine/js/lookup/a9.js:114]: Done in https://github.com/Wikia/app/pull/13603/commits/2d523760307796875dadae429ac8bc34e9f5533b

File: extensions/wikia/AdEngine/js/lookup/a9.js (hunk 8 above)
Comment (fraszczakszymon) [reply extensions/wikia/AdEngine/js/lookup/a9.js:132]: :+1:

File: extensions/wikia/AdEngine/js/lookup/a9.js (hunk 9 above)
Comment (fraszczakszymon) [reply extensions/wikia/AdEngine/js/lookup/a9.js:62]: :+1:

**Pull request merged**